src/bin/str.rs

```diff
@@ -89,88 +132,92 @@ define_language! {
         "+" = Add([Id; 2]),
         "-" = Sub([Id; 2]),
         Lit(Constant),
-        Var(egg::Symbol),
     }
 }
 
-/// Concatenate strings.
-fn concat(a: &[u8], b: &[u8]) -> SmallStr {
-    a.iter().chain(b).copied().collect()
-}
-
 impl SynthLanguage for Lang {
     type Constant = Constant;
     type Type = Type;
 
     fn get_type(&self) -> Self::Type {
-        Type::Top
+        match self {
+            Lang::Lit(c) => match c {
+                Constant::Num(_) => Type::Number,
+                Constant::Str(_) => Type::String,
+            },
+            Lang::SVar(_)
+            | Lang::Concat(_)
+            | Lang::Replace(_)
+            | Lang::At(_)
+            | Lang::SubStr(_)
+            | Lang::FromInt(_) => Type::String,
+            Lang::NVar(_)
```

Comment:
ah yeah `str` is an interesting domain because `get_type` is not all the same for the ops! Cool cool.